pkg/controllers/management/clusterprovisioner/provisioner.go

```diff
@@ -214,6 +241,39 @@ func (p *Provisioner) Updated(cluster *v3.Cluster) (runtime.Object, error) {
 	return obj.(*v3.Cluster), err
 }
 
+func (p *Provisioner) startCleanUpHelmReleases(cluster *v3.Cluster) (*v3.Cluster, error) {
+	object, err := apimgmtv3.ClusterConditionHelmReleasesMigrated.Do(cluster, func() (runtime.Object, error) {
+		err := p.cleanupHelmReleases(cluster)
+		if err != nil {
+			return cluster, err
+		}
+
+		cluster, err = p.Clusters.Get(cluster.Name, metav1.GetOptions{})
+		if err != nil {
+			// if we failed to get a cluster, something bad must have happened
+			return cluster, errors.Wrapf(err, "[updated] failed to get updated cluster %#v after startCleanUpHelmReleases", cluster.Name)
+		}
+
+		return cluster, nil
+	})
```

Comment:
I'm a bit nervous about returning here, but I'm not sure how the `Do` function of the condition works. Do we expect the Do function to update the status to True only if the function succeeds? 

I'm concerned that in the case where we get an error, you could end up updating the cluster with a resolved condition but returning an error. Since the condition is updated, the next run won't attempt to re-clean the releases, which I think could be bad.

Reply:
@rmweir could you weigh on this?
I don't have enough context on how `.Do` works to know how would this scenario go.

Reply:
If you return an error the Do function will set the condition to False with an error message. You can view the logic [here](https://github.com/rancher/norman/blob/master/condition/condition.go#L227).